### PR TITLE
RATIS-1120. Terminate election in advance if majority peers reject

### PR DESCRIPTION
**Pull request by dengziming**
## What changes were proposed in this pull request?

In the leader election approach, the candidate will change to leader in advance if the majority voters vote for it. we can also terminate the approach if majority voters reject it.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1120

## How was this patch tested?

unit test.

### CODE REVIEW

**Comment by runzhiwang** [comment]
@dengziming Thanks for the first PR.  Please fill What changes were proposed in this pull request?,  What is the link to the Apache JIRA?, How was this patch tested?.

**Comment by dengziming** [comment]
@runzhiwang ,Hi, I changed the comment, please take a look.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/PeerConfiguration.java [hunk 1]
```diff
@@ -94,6 +94,16 @@ boolean hasMajority(Collection<RaftPeerId> others, RaftPeerId selfId) {
     return num > size() / 2;
   }
 
+  boolean majorityImpossibleWithout(Collection<RaftPeerId> rejected) {
```
Comment (runzhiwang) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/PeerConfiguration.java:97]: Do you think rename majorityImpossibleWithout to majorityRejectVote maybe better ?

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftConfiguration.java [hunk 2]
```diff
@@ -206,6 +206,12 @@ boolean hasMajority(Collection<RaftPeerId> others, RaftPeerId selfId) {
         (oldConf == null || oldConf.hasMajority(others, selfId));
   }
 
+  /** @return true if the rejects are in the half(impossible for others in majority). */
+  boolean majorityImpossibleWithout(Collection<RaftPeerId> rejects) {
+    return conf.majorityImpossibleWithout(rejects) &&
+            (oldConf == null || oldConf.majorityImpossibleWithout(rejects));
+  }
```
Comment (runzhiwang) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/RaftConfiguration.java:213]: I think this check it not totally reverse  of hasMajority.  The reverse of hasMajority is `!conf.hasMajority(others, selfId) || !(oldConf == null || oldConf.hasMajority(others, selfId))` i.e.` !conf.hasMajority(others, selfId) || (oldConf != null && !oldConf.hasMajority(others, selfId))``. i.e. ` conf.majorityImpossibleWithout(rejects) || (oldConf != null && !oldConf.majorityImpossibleWithout(rejects))``

File: ratis-server/src/main/java/org/apache/ratis/server/impl/PeerConfiguration.java (hunk 1 above)
Comment (dengziming) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/PeerConfiguration.java:97]: OK, firstly I assume rejected votes is not necessarily a majority, maybe half is enough. But I think majorityRejectVote is more self-explanatory even it is not very accurate.

Comment (runzhiwang) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/PeerConfiguration.java:97]: If you have better name, I also can accept it.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/RaftConfiguration.java (hunk 2 above)
Comment (dengziming) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/RaftConfiguration.java:213]: This is right! I just forgot this.

**Comment by runzhiwang** [comment]
reopen pr to trigger ci.

**Comment by runzhiwang** [comment]
@dengziming Thanks the first patch. I have merged it. Looking forward to your more contribution.

**Pull request merged**